### PR TITLE
ipc4: add support for voice recorder in windows 10

### PR DESCRIPTION
**Pull request by RanderWang**
Voice recorder can capture voice and save it to hard disk. This patch fixes some bugs and add support of format conversion used by it.

### CODE REVIEW

File: src/include/sof/audio/pcm_converter.h [hunk 1]
```diff
@@ -103,6 +103,7 @@ struct pcm_func_vc_map {
 	enum sof_ipc_frame sink;	/**< sink frame container format */
 	enum sof_ipc_frame valid_sink_bits;	/**< sink frame format */
 	uint32_t type;	/**< gateway type */
+	uint32_t direction;	/**< support playback, capture or both */
```
Comment (lgirdwood) [inline src/include/sof/audio/pcm_converter.h:106]: use the enum ?

**Review by lgirdwood — CHANGES_REQUESTED** [review]
@marcinszkudlinski @tmleman @rfredzim any comments ?

File: src/ipc/ipc4/handler.c [hunk 2]
```diff
@@ -379,12 +380,46 @@ static int ipc_wait_for_compound_msg(void)
...
+	ipc = ipc_get();
+
+	/* only dai has direction based on gateway type */
+	dai = pipeline_get_dai_comp(ppl_id[0], &latency);
+	if (!dai) {
+		tr_err(&ipc_tr, "can't find dai comp");
+		return -EINVAL;
+	}
+
+	list_for_item(clist, &ipc->comp_list) {
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/ipc/ipc4/handler.c:400]: sure

File: src/ipc/ipc4/handler.c [hunk 3]
```diff
@@ -379,12 +380,46 @@ static int ipc_wait_for_compound_msg(void)
 	return ret;
 }
 
+static int update_dir_to_pipeline_component(uint32_t *ppl_id, uint32_t count)
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/ipc/ipc4/handler.c:383]: sure

File: src/include/sof/audio/pcm_converter.h (hunk 1 above)
Comment (RanderWang) [reply src/include/sof/audio/pcm_converter.h]: sure

File: src/ipc/ipc4/handler.c [hunk 4]
```diff
@@ -379,12 +380,54 @@ static int ipc_wait_for_compound_msg(void)
...
+	/* set dai direction to the component in the pipline array */
+	list_for_item(clist, &ipc->comp_list) {
+		icd = container_of(clist, struct ipc_comp_dev, list);
+		if (icd->type != COMP_TYPE_COMPONENT)
+			continue;
+
+		for (i = 0; i < count; i++) {
+			if (ipc_comp_pipe_id(icd) == ppl_id[i]) {
+				icd->cd->direction = dai->direction;
+				break;
```
Comment (tmleman) [inline src/ipc/ipc4/handler.c:416]: How it will behave when hda copiers, both host output and input, are under the same pipeline?

**Review by tmleman — CHANGES_REQUESTED** [review]

File: src/ipc/ipc4/handler.c (hunk 4 above)
Comment (RanderWang) [reply src/ipc/ipc4/handler.c:416]: @tmleman  good point. And do we have a use case that dai copier to dai copier ?

Comment (tmleman) [reply src/ipc/ipc4/handler.c:416]: @RanderWang yes, mainly for validation.

Comment (RanderWang) [reply src/ipc/ipc4/handler.c:416]: Thanks, I will redesign it

**Comment by RanderWang** [comment]
@tmleman  please test this PR, I updated it . thanks

**Review by tmleman — APPROVED** [review]

**Pull request merged**